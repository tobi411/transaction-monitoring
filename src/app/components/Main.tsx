'use client';
import { useEffect, useMemo, useState } from 'react';
import { AccountResponse, API_STATUS } from '../lib/types';
import Dropdown from './Dropdown';
import Account from './Account';
import axios from 'axios';
import Banner from './Banner';

const ACCOUNT_STORAGE_KEY = 'selectedAccount';

export default function Main() {
  const [accounts, setAccounts] = useState<Array<AccountResponse>>([]);
  const [selectedAccountId, setSelectedAccountId] = useState<string>();
  const [fetchAccountStatus, setFetchAccountState] = useState<API_STATUS>();

  const selectedAccountInStorage = localStorage.getItem(ACCOUNT_STORAGE_KEY);

  const getAccounts = async () => {
    try {
      setFetchAccountState(API_STATUS.LOADING);
      const response = await axios.get(
        'https://paloma-financial-auditor-0aff70148dbe.herokuapp.com/accounts'
      );
      const accountsResponse = response.data.data;
      setFetchAccountState(API_STATUS.COMPLETED);
      setAccounts(accountsResponse);
    } catch (e) {
      setFetchAccountState(API_STATUS.FAILED);
    }
  };

  useEffect(() => {
    getAccounts();
  }, []);

  const selectedAccount = useMemo(() => {
    if (!selectedAccountId || accounts?.length === 0) return undefined;
    return accounts?.find((account) => account.accountId === selectedAccountId);
  }, [selectedAccountId, accounts]);

  useEffect(() => {
    if (selectedAccountInStorage) {
      setSelectedAccountId(selectedAccountInStorage);
    }
  }, [selectedAccountInStorage]);

  const dropdownOptions = useMemo(() => {
    return accounts.map((account) => ({
      label: account.accountName,
      value: account.accountId,
      id: account.accountId,
    }));
  }, [accounts]);

  const handleAccountSelected = (value: string) => {
    setSelectedAccountId(value);
    localStorage.setItem(ACCOUNT_STORAGE_KEY, value);
  };

  return (
    <>
      {fetchAccountStatus === API_STATUS.FAILED && <Banner />}
      <div className='container mx-auto py-6'>
        <div className='flex'>
          <Dropdown
            label='Account'
            options={dropdownOptions}
            value={selectedAccount?.accountName}
            onClick={(value) => handleAccountSelected(value)}
          />
          <Account account={selectedAccount} />
        </div>
      </div>
    </>
  );
}
