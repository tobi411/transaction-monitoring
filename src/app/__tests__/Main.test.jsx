import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Main from '../components/Main';

jest.mock('axios');

describe('Page', () => {
  it('render content', async () => {
    await waitFor(() => {
      render(<Main accounts={[]} />);
    });

    const content = screen.getByText('Account');

    expect(content).toBeInTheDocument();
  });

  it('matches previous snapshot', () => {
    let domTree = renderer.create(<Main />);
    expect(domTree).toMatchSnapshot();
  });
});
