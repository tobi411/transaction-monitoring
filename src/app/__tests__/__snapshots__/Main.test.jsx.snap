// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page matches previous snapshot 1`] = `
<div
  className="container mx-auto py-6"
>
  <div
    className="flex"
  >
    <div
      className="relative inline-block text-left"
      data-headlessui-state=""
    >
      <div
        className="w-52"
      >
        <label
          className="font-bold text-sm"
          htmlFor="dropdown-button"
        >
          Account
        </label>
        <button
          aria-expanded={false}
          aria-haspopup="menu"
          autoFocus={false}
          className="inline-flex w-full justify-between gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm text-gray-900  shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
          data-headlessui-state=""
          id="dropdown-button"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPointerDown={[Function]}
          onPointerUp={[Function]}
          onTouchStart={[Function]}
          type="button"
        >
          <div />
          <svg
            aria-hidden="true"
            className="-mr-1 size-5 text-gray-400 "
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M5.22 8.22a.75.75 0 0 1 1.06 0L10 11.94l3.72-3.72a.75.75 0 1 1 1.06 1.06l-4.25 4.25a.75.75 0 0 1-1.06 0L5.22 9.28a.75.75 0 0 1 0-1.06Z"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
